File: temporal-research/08_template_runner/interpreter.py
"""
Generic workflow interpreter: executes any Template definition.

Execution model
───────────────
Every node is launched as a concurrent coroutine immediately at workflow start.
Each coroutine internally waits (via workflow.wait_condition) until ALL its
declared dependencies reach a terminal state (completed / skipped / failed).

Once dependencies are terminal:
  1. Evaluate condition  → false or any dep was skipped/failed → SKIPPED
  2. Resolve ${} param references against context
  3. Dispatch the activity by name → RUNNING → COMPLETED or FAILED

asyncio.gather across all nodes gives maximum concurrency.
workflow.wait_condition is durable — the timer survives worker restarts.
"""
import asyncio
from datetime import timedelta
from typing import Any, Optional

from temporalio import workflow
from temporalio.exceptions import ApplicationError

with workflow.unsafe.imports_allowed():
    from models import NodeState, NodeStatus, Template, TemplateNode


_TIMEOUT  = timedelta(seconds=120)
_TERMINAL = frozenset({NodeStatus.COMPLETED, NodeStatus.SKIPPED, NodeStatus.FAILED})


# ── Param resolution ──────────────────────────────────────────────────────────

def _resolve(params: dict[str, Any], context: dict) -> dict:
    """
    Replace ${source.field} tokens with values from the execution context.

      ${ctx.region}          → context["_input"]["region"]
      ${subnet_a.subnet_id}  → context["subnet_a"]["subnet_id"]
    """
    out: dict[str, Any] = {}
    for key, value in params.items():
        if isinstance(value, str) and value.startswith("${") and value.endswith("}"):
            ref = value[2:-1]
            parts = ref.split(".", 1)
            src_key, field = parts[0], (parts[1] if len(parts) > 1 else None)
            src = context.get("_input", {}) if src_key == "ctx" else context.get(src_key, {})
            out[key] = src.get(field) if (field and isinstance(src, dict)) else src
        else:
            out[key] = value
    return out


# ── Interpreter workflow ──────────────────────────────────────────────────────

@workflow.defn
class TemplateRunnerWorkflow:

    def __init__(self) -> None:
        self._states: dict[str, NodeState] = {}
        self._cancelled = False

    # ── External interface ─────────────────────────────────────────────────

    @workflow.signal
    async def cancel(self) -> None:
        """Soft-cancel: marks all pending/running nodes as SKIPPED."""
        workflow.logger.info("Cancellation requested")
        self._cancelled = True

    @workflow.query
    def get_execution_graph(self) -> dict:
        """
        Returns the full node state map — including PENDING and SKIPPED nodes.
        This is the key difference from the Temporal event history: every
        declared node is always present, regardless of whether it ran.
        """
        return {
            "nodes": [
                {
                    "id":          s.node_id,
                    "name":        s.node_name,
                    "condition":   s.condition,
                    "status":      s.status.value,
                    "output":      s.output,
                    "error":       s.error,
                    "skip_reason": s.skip_reason,
                }
                for s in self._states.values()
            ]
        }

    # ── Entry point ────────────────────────────────────────────────────────

    @workflow.run
    async def run(self, template: Template, input_context: dict) -> dict:
        # Initialise every node as PENDING so the query returns the full graph
        # immediately — before any activity runs.
        for node in template.nodes:
            self._states[node.id] = NodeState(
                node_id=node.id,
                node_name=node.name,
                condition=node.condition,
            )

        # Embed input context so ${ctx.field} resolution works
        context: dict[str, Any] = {"_input": input_context}

        # Launch ALL nodes concurrently.  Each waits internally for its deps.
        await asyncio.gather(*[
            self._run_node(node, context) for node in template.nodes
        ])

        counts = {s.value: 0 for s in NodeStatus}
        for state in self._states.values():
            counts[state.status.value] += 1
        return counts

    # ── Per-node coroutine ─────────────────────────────────────────────────

    async def _run_node(self, node: TemplateNode, context: dict) -> None:
        # Step 1: wait until every declared dependency is terminal.
        # workflow.wait_condition is durable — the predicate is re-evaluated
        # after each activity / signal event, not on a timer.
        if node.depends_on:
            await workflow.wait_condition(
                lambda: all(
                    self._states[dep].status in _TERMINAL
                    for dep in node.depends_on
                )
            )

        if self._cancelled:
            self._set_skipped(node.id, "workflow cancelled")
            return

        # Step 2: decide whether to skip or execute.
        skip, reason = self._skip_reason(node, context)
        if skip:
            self._set_skipped(node.id, reason)
            return

        # Step 3: execute the activity.
        await self._execute(node, context)

    async def _execute(self, node: TemplateNode, context: dict) -> None:
        state = self._states[node.id]
        state.status = NodeStatus.RUNNING

        resolved = _resolve(node.params, context)

        try:
            output = await workflow.execute_activity(
                node.activity,          # dispatched by string name
                resolved,               # single dict arg
                start_to_close_timeout=_TIMEOUT,
            )
            state.status = NodeStatus.COMPLETED
            state.output = output if isinstance(output, dict) else {"result": output}
            # Publish output into context so downstream nodes can reference it
            context[node.id] = state.output
        except Exception as exc:
            state.status = NodeStatus.FAILED
            state.error   = str(exc)
            raise

    # ── Helpers ────────────────────────────────────────────────────────────

    def _skip_reason(self, node: TemplateNode, context: dict) -> tuple[bool, str]:
        # Explicit condition
        if node.condition is not None:
            if not context.get("_input", {}).get(node.condition):
                return True, f"condition '{node.condition}' = false"

        # Cascade: skip if any dependency was skipped or failed
        for dep_id in node.depends_on:
            dep = self._states.get(dep_id)
            if dep and dep.status == NodeStatus.SKIPPED:
                return True, f"upstream '{dep_id}' was skipped"
            if dep and dep.status == NodeStatus.FAILED:
                return True, f"upstream '{dep_id}' failed"

        return False, ""

    def _set_skipped(self, node_id: str, reason: str) -> None:
        s = self._states[node_id]
        s.status      = NodeStatus.SKIPPED
        s.skip_reason = reason

File: temporal-research-go/08_template_runner/interpreter.go
package template_runner

import (
	"fmt"
	"strings"
	"time"

	"go.temporal.io/sdk/workflow"
)

// isTerminal returns true when a node has reached a final (non-running) state.
func isTerminal(s NodeStatus) bool {
	return s == Completed || s == Skipped || s == Failed
}

// isFalsy checks whether a value from the input context should be treated as false/absent.
func isFalsy(v any) bool {
	if v == nil {
		return true
	}
	switch val := v.(type) {
	case bool:
		return !val
	case string:
		return val == "" || val == "false"
	case int:
		return val == 0
	case float64:
		return val == 0
	}
	return false
}

// resolveValue replaces a single "${ctx.field}" or "${node_id.field}" reference with its value.
// If the string is not a reference pattern it is returned unchanged.
func resolveValue(raw string, execContext map[string]any) any {
	if !strings.HasPrefix(raw, "${") || !strings.HasSuffix(raw, "}") {
		return raw
	}
	inner := raw[2 : len(raw)-1] // strip ${ and }
	dot := strings.Index(inner, ".")
	if dot < 0 {
		return raw
	}
	source := inner[:dot]
	field := inner[dot+1:]

	if source == "ctx" {
		if inputCtx, ok := execContext["_input"]; ok {
			if m, ok := inputCtx.(map[string]any); ok {
				if v, exists := m[field]; exists {
					return v
				}
			}
		}
		return nil
	}

	// node reference
	if nodeOut, ok := execContext[source]; ok {
		if m, ok := nodeOut.(map[string]any); ok {
			if v, exists := m[field]; exists {
				return v
			}
		}
	}
	return nil
}

// resolveParams replaces all parameter values that are reference strings with their resolved values.
func resolveParams(params map[string]any, execContext map[string]any) map[string]any {
	if params == nil {
		return map[string]any{}
	}
	resolved := make(map[string]any, len(params))
	for k, v := range params {
		if s, ok := v.(string); ok {
			resolved[k] = resolveValue(s, execContext)
		} else {
			resolved[k] = v
		}
	}
	return resolved
}

// snapshotGraph creates a copy of the current execution state for query handlers.
func snapshotGraph(templateID string, states map[string]*NodeState) ExecutionGraph {
	nodes := make(map[string]NodeState, len(states))
	for id, ns := range states {
		nodes[id] = *ns
	}
	return ExecutionGraph{
		TemplateID: templateID,
		Nodes:      nodes,
	}
}

// TemplateRunnerWorkflow interprets a Template DAG, executing each node as an activity
// while respecting dependency ordering, skip conditions, and cascading skips/failures.
func TemplateRunnerWorkflow(ctx workflow.Context, tmpl Template, inputCtx map[string]any) (map[string]any, error) {
	logger := workflow.GetLogger(ctx)

	// 1. Initialise all node states as Pending.
	states := make(map[string]*NodeState, len(tmpl.Nodes))
	for _, node := range tmpl.Nodes {
		states[node.ID] = &NodeState{
			NodeID:   node.ID,
			NodeName: node.Name,
			Status:   Pending,
		}
		if node.Condition != "" {
			states[node.ID].Condition = node.Condition
		}
	}

	// 2. Register the query handler so callers can inspect progress at any time.
	if err := workflow.SetQueryHandler(ctx, "get_execution_graph", func() (ExecutionGraph, error) {
		return snapshotGraph(tmpl.ID, states), nil
	}); err != nil {
		return nil, fmt.Errorf("failed to register query handler: %w", err)
	}

	// 3. Shared execution context: "_input" holds the caller-supplied context;
	//    completed node outputs are stored under their node ID.
	execContext := map[string]any{
		"_input": inputCtx,
	}

	// 5. Done channel — one token per node goroutine.
	doneCh := workflow.NewBufferedChannel(ctx, len(tmpl.Nodes))

	// 6. Fan-out: launch one coroutine per node.
	for _, node := range tmpl.Nodes {
		node := node // capture loop variable
		workflow.Go(ctx, func(gCtx workflow.Context) {
			state := states[node.ID]

			// a. Wait until every dependency has reached a terminal state.
			if len(node.DependsOn) > 0 {
				workflow.Await(gCtx, func() bool {
					for _, depID := range node.DependsOn {
						if dep, ok := states[depID]; !ok || !isTerminal(dep.Status) {
							return false
						}
					}
					return true
				})
			}

			// b. Determine whether this node should be skipped.
			skipReason := ""

			// Condition check
			if node.Condition != "" {
				val, exists := inputCtx[node.Condition]
				if !exists || isFalsy(val) {
					skipReason = fmt.Sprintf("condition %q is false or missing", node.Condition)
				}
			}

			// Cascade skip/fail from dependencies
			if skipReason == "" {
				for _, depID := range node.DependsOn {
					dep := states[depID]
					if dep.Status == Skipped {
						skipReason = fmt.Sprintf("dependency %q was skipped", depID)
						break
					}
					if dep.Status == Failed {
						skipReason = fmt.Sprintf("dependency %q failed", depID)
						break
					}
				}
			}

			if skipReason != "" {
				state.Status = Skipped
				state.SkipReason = skipReason
				logger.Info("Node skipped", "nodeID", node.ID, "reason", skipReason)
				doneCh.Send(gCtx, struct{}{})
				return
			}

			// c. Mark as running.
			state.Status = Running
			logger.Info("Node running", "nodeID", node.ID, "activity", node.Activity)

			// d. Resolve parameter references.
			resolvedParams := resolveParams(node.Params, execContext)

			// e. Execute the activity.
			ao := workflow.ActivityOptions{
				StartToCloseTimeout: 30 * time.Second,
			}
			actCtx := workflow.WithActivityOptions(gCtx, ao)

			var output map[string]any
			err := workflow.ExecuteActivity(actCtx, node.Activity, resolvedParams).Get(actCtx, &output)

			// f. Record outcome.
			if err != nil {
				state.Status = Failed
				state.Error = err.Error()
				logger.Info("Node failed", "nodeID", node.ID, "error", err)
			} else {
				state.Status = Completed
				state.Output = output
				execContext[node.ID] = output
				logger.Info("Node completed", "nodeID", node.ID)
			}

			doneCh.Send(gCtx, struct{}{})
		})
	}

	// 7. Wait for all goroutines to finish.
	for i := 0; i < len(tmpl.Nodes); i++ {
		doneCh.Receive(ctx, nil)
	}

	// 8. Build and return the final execution context (minus the internal "_input" key).
	result := make(map[string]any, len(execContext))
	for k, v := range execContext {
		if k != "_input" {
			result[k] = v
		}
	}
	return result, nil
}
